add middleware for request prioritization

bohde opened the pull request:
This adds a middleware for overload protection, that is intended to help protect against malicious scrapers. It does this via  [`codel`](https://github.com/bohde/codel), which will perform the following:

1. Limit the number of in-flight requests to some user defined max
2. When in-flight requests have reached their max, begin queuing requests, with logged in requests having priority above logged out requests
3. Once a request has been queued for too long, it has a percentage chance to be rejected based upon how overloaded the entire system is.

When a server experiences more traffic than it  can handle, this has the effect of keeping latency low for logged in users, while rejecting just enough requests from logged out users to keep the service from being overloaded.

Below are benchmarks showing a system at 100% capacity and 200% capacity in a few different configurations. The 200% capacity is shown to highlight an extreme case. I used [hey](https://github.com/rakyll/hey) to simulate the bot traffic:

```
hey -z 1m -c 10 "http://localhost:3000/rowan/demo/issues?state=open&type=all&labels=&milestone=0&project=0&assignee=0&poster=0&q=fix"
```

The concurrency of 10 was chosen from experiments where my local server began to experience higher latency.

Results

| Method | Concurrency |  p95 latency | Successful RPS | Requests Dropped | 
|--------|--------|--------|--------|--------|
| QoS Off | 10 | 0.2960s | 44 rps | 0% |
| QoS Off | 20 | 0.5667s | 44 rps | 0%|
| QoS On | 20 |  0.4409s | 48 rps | 10% |
| QoS On 50% Logged In* | 10 | 0.3891s | 33 rps | 7% | 
| QoS On 50% Logged Out* | 10  | 2.0388s | 13 rps | 6% |

Logged in users were given the additional parameter ` -H "Cookie: i_like_gitea=<session>`.

Tests with `*` were run at the same time, representing a workload with mixed logged in & logged out users. Results are separated to show prioritization, and how logged in users experience a 100ms latency increase under load, compared to the 1.6 seconds logged out users see.

wxiaoguang commented on routers/common/qos.go:
```diff
@@ -0,0 +1,121 @@
+}
+
+func IsGitContents(path string) bool {
+	parts := []string{
+		"refs",
+		"archive",
+		"commit",
+		"graph",
+		"blame",
+		"branches",
+		"tags",
+		"labels",
+		"stars",
+		"search",
+		"activity",
+		"wiki",
+		"watchers",
+		"compare",
+		"raw",
+		"src",
+		"commits",
+	}
```
I do not think such "guessing" could work well .....

bohde replied:
When I've developed similar middlewares, I've also provided middlewares to classify a route as having a given priority, by attaching metadata to the individual routes. This approach would avoid the guessing nature, but I'm not quite sure how to do it with the middleware structure here.

wxiaoguang commented:
> Results ....

TBH, according to your test result, I do not see it is really useful .......

a1012112796 reviewed (commented):
I think at first step, just distinguish between login and not is enough. and looks it can be added for the api router also.

bohde commented:
> TBH, according to your test result, I do not see it is really useful .......

We've been running this patch for awhile, and while it takes a bit of tuning to get the right settings for a given server it has alleviated a lot of outage concerns from waves of scrapers that don't respect robots.txt.

> I think at first step, just distinguish between login and not is enough. and looks it can be added for the api router also.

This was in my initial version we ran, but then users who are attempting to login can still experience lower priority traffic. Since the major source of traffic for public code hosting are scrapers trying to download content, deprioritizing those routes mitigates that concern.

wxiaoguang commented:
> > TBH, according to your test result, I do not see it is really useful .......
> 
> We've been running this patch for awhile, and while it takes a bit of tuning to get the right settings for a given server it has alleviated a lot of outage concerns from waves of scrapers that don't respect robots.txt.

So the original intention is to fight with crawlers/spiders/robots?

wxiaoguang commented:
> Yes, I mentioned this in my description

Sorry, missed that part. The "perform the following" part and "result" part attracted my attention .....


If the intention is to "protect against malicious scrapers", I think the `parts := []string{`  could be improved, a similar function like `ParseGiteaSiteURL` could help the route path handling. And I think we need some tests to cover the new middleware's behavior.

wxiaoguang commented:
I think I have some ideas about how to handle these "route paths" clearly. Could I push some commits into your branch?

bohde commented:
> I think I have some ideas about how to handle these "route paths" clearly. Could I push some commits into your branch?

Could you send a PR to my branch, or sketch it out in another commit? This would prevent merge conflicts on my branch

wxiaoguang commented:
> sketch it out in another commit

Sorry, don't quite understand what it exactly means .... 

> Could you send a PR to my branch

* https://github.com/bohde/gitea/pull/1

We can use chi router's "RoutePattern", and make code testable.

bohde commented:
> update: Close since there is no interest after 2 weeks

Just getting back to this now as I was dealing with other priorities. I could incorporate the above changes, but the near complete rewrite of my PR makes it difficult to isolate the changes you are asking, and makes it incompatible with the production implementation we are running. This makes it difficult to test these changes against the live traffic we are seeing to ensure they still perform well.

bohde commented:
I applied review feedback, and synthesized into something that is easily backportable to 1.23 by avoiding APIs added in `main`. @wxiaoguang's PR did miss some important endpoints we sometimes see heavy traffic on since it didn't check every possible delimited portion of the subpath against the string in my original implementation.  However, because the `RoutePattern` can be pulled directly in this middleware using @wxiaoguang's method, this can all be simplified to the following priority scheme:

| Logged In | Repo Context | Priority |
|--------|--------|--------|
| Yes | - | High |
| No | Yes | Low |
| No | No | Default | 

This allows key flows such as login and explore to still be prioritized well, with only the repo specific endpoints likely to be throttled. In my experience, this catches the vast majority of what scrapers are targeting.

bohde commented on routers/common/qos.go:
```diff
@@ -0,0 +1,112 @@
+
+	// If we're operating in the context of a repo, assign low priority
+	routePattern := rctx.RoutePattern()
+	if strings.HasPrefix(routePattern, "/{username}/{reponame}") {
```
Is it the technique you object to? If so, that is why I made it a configurable setting that defaults to off. 

In practice, this only returns enough 503s to prevent the server from being overloaded, allowing the maximum amount of traffic through while prioritizing logged in users. The alternative is to have a slow or unavailable server for all users when hit with a large amount of crawler traffic.

wxiaoguang replied:
> Is it the technique you object to? If so, that is why I made it a configurable setting that defaults to off.

The discussion is based on when the configurable setting that defaults to on: whether it is good enough when the protection is turned on.

According to the feedbacks of "Setting to disable expensive endpoints for anonymous users #33966", it's pretty safe to expose the home pages and some other entry pages.

wxiaoguang replied:
1. There are different things.

* `repo.Home` is the handler for all "repo code" request, include '/src/....`. What you have seen are from various `/src/` requests.
* But for a repo's landing page "home", `/owner/repo` (no more path), it is only a single endpoint and pretty safe to crawlers.

2. I think it's better to redirect to the user login page instead of returning 503.

wxiaoguang replied:
And according some feedbacks from `https://github.com/go-gitea/gitea/issues/33407#issuecomment-2776671993` , IMO a complete design could be like this:

1. Anti-crawler policy: block expensive pages, apply QoS to expensive pages
2. anti-crawler action: respond 503, redirect to login, show a JS page to do some proof-of-work or use some rate limit token (this one seems to provide best user experience).

Then end users could choose a combination to satisfy their requirements.

bohde replied:
> But for a repo's landing page "home", /owner/repo (no more path), it is only a single endpoint and pretty safe to crawlers.

Adjusted this in https://github.com/go-gitea/gitea/pull/33951/commits/dac63a77d8e091a4e9fff1354ef08b71fce52c33 to assign the repo home page. 

> I think it's better to redirect to the user login page instead of returning 503.

I disagree with that design. Some instances may set [`service.DISABLE_REGISTRATION` ](https://docs.gitea.com/next/administration/config-cheat-sheet#service-service) to `true` but still allow for public repos. In this case, redirecting a non registered user is a confusing user experience. This is especially confusing due to the probabilistic nature of load shedding: sometimes a user may be able to successfully view a commit for instance, and sometimes they are redirected to a login page. 

HTTP 503 is the semantic status code for a temporary, retryable error that is not the fault of the user. By returning this, we also allow other server operators to parse their logs and generate metrics and alerts when load shedding is triggering too much. 

> IMO a complete design could be like this:

This isn't intended  complete design for crawlers, but helps address crawlers by reducing  the impact of they have on other traffic and simplifying capacity planning for server operators. The way I think about it, is that the server can then be sized to handle all high and default priority traffic, with a buffer to handle the low priority traffic. In the event that traffic begins to overwhelm the server, it is then very likely that the only traffic that will be shed is low priority, while high and default priority traffic continues to be served quickly. This then allows the operator to make other decisions on whether or not to adjust resources for the server, add other bot detection and rejection mechanisms, or wait for the situation to resolve. Either way, the prevention of a full outage buys time for the operator to make these decisions without the pressure of dealing with a large scale incident. 

We also use additional infrastructure level mitigations to help control crawlers, but those are not Gitea specific.

wxiaoguang replied:
Thank you very much for the explanation.

Actually I don't understand why "real users should see 503 error when the site is being crawled".

And since I don't run public Gitea instance, so I don't have a comprehensive understanding of this issue. Public site admins (eg: gitea.com) could have comprehensive understanding and help.

bohde replied:
> Actually I don't understand why "real anonymous users should see 503 error when the site is being crawled".

Part of my recommendation for this is simplicity in error handling. By serving a 503, I don't need to worry about the client then immediately issuing another request to the login handler, avoiding adding more load to the server. 

The prioritization rules here help shift the probabilities so that it is more likely that any traffic that is shed is traffic from a crawler, making it even less likely that an anonymous user experiences it. Even in the case where an anonymous user does experience a 503, the alternative is no traffic shedding which could lead to a full outage of the server and an even worse experience.

wxiaoguang replied:
My proposal is:

https://github.com/go-gitea/gitea/pull/33951#discussion_r2032324964

> 1. ....
> 2. ... show a JS page to do some proof-of-work or use some rate limit token (this one seems to provide best user experience).
> Then end users could choose a combination to satisfy their requirements.

It should be able to provide best user experiences for public instances. These feedbacks are from the comments I shared above.

(But I can't comment more since I don't run a public instance)

bohde replied:
I'm going to mark this conversation as resolve then, and request reviews from other maintainers. For context, we do run a public instance and the code here has addressed a significant amount of the issues we see with crawlers, with little to no impact negative impact on users. As a whole, this is a net positive for them, since latency has remained low despite crawler traffic increasing.

wxiaoguang replied:
No problem, I will quote it in a separate comment to let maintainers decide (I am also interested in how to handle this edge case for public instances)

lunny commented:
Alternatively, a configuration item may be introduced to return a 503 status code or redirect users to a custom URL.

bohde commented:
> Alternatively, a configuration item may be introduced to return a 503 status code or redirect users to a custom URL.

The fundamental problem here is that if we send a redirect instead of a 503, we now generate more load when the user follows the redirect. This causes us to generate more load, instead of removing load. This can cause feedback loops if this newly added load also generates more redirects.

bohde commented:
Just to validate this, I prototyped the redirect method locally and ran benchmarks using `hey`, which follows redirects by default. The command I ran was:

```
hey -z 1m -c 20 "http://localhost:3000/rowan/demo/issues?state=open&type=all&labels=&milestone=0&project=0&assignee=0&poster=0&q=fix"
```

The results are below, but the redirect method shows worse performance on p95 latency than just not having QoS turned on, because of the added load redirects introduce. It does however show higher RPS (and mean latency), because rendering a login page is less intensive than the issue search. 

| Method | Concurrency |  p95 latency | Successful RPS | Requests Dropped | 
|--------|--------|--------|--------|--------|
| QoS Off | 20 | 0.4912s | 50 rps | 0%|
| QoS On (503) | 20 |  0.4058s | 63 rps | 8% |
| QoS On (Redirect) | 20 | 0.8329s | 60 rps | 0% |

silverwind commented:
I recommend using status code 429 (Too Many Requests) instead of 503 (Service Unavailable) for any rate-limited responses. The intend is more clear with that code.

bohde commented:
> I recommend using status code 429 (Too Many Requests) instead of 503 (Service Unavailable) for any rate-limited responses. The intent is more clear with that code and it blames the client, not the server.

This isn't quite the right semantics. [429 Too Many Requests](https://www.rfc-editor.org/rfc/rfc6585#section-4) means "that the user has sent too many requests in a given amount of time", which is not necessarily the case here. For example, a user can on their first request to the server have that request rejected by this middleware. This is not an issue with the client, but the server, in that it has determined it cannot service the request. 

> For bonus points, one could also send the [`Retry-After` header](https://developer.mozilla.org/en-US/docs/Web/HTTP/Reference/Headers/Retry-After) to indicate when further requests may be accepted.

This would also apply to [503 Service Unavailable](https://www.rfc-editor.org/rfc/rfc9110#name-503-service-unavailable), but any value here is not likely to have a positive effect. In the case of load shedding like this, that field is useful when the clients are cooperative with the server, willingly backing off for that time period. However, crawlers are not cooperative (e.g. [they do not respect robots.txt](https://github.com/go-gitea/gitea/issues/33966)), so it is unlikely they would respect this field. 

Hypothetically, an operator of a public instance could implement this field and see how it affects crawlers, but the result would vary from crawler to crawler.

wxiaoguang commented:
> > Alternatively, a configuration item may be introduced to return a 503 status code or redirect users to a custom URL.
> 
> The fundamental problem here is that if we send a redirect instead of a 503, we now generate more load when the user follows the redirect. This causes us to generate more load, instead of removing load. This can cause feedback loops if this newly added load also generates more redirects.

That's not true according to the feedbacks of the existing "REQUIRE_SIGNIN_VIEW=expensive" config option. All users said that there is no more CPU load.

wxiaoguang commented:
> The results are below, but the redirect method shows worse performance on p95 latency than just not having QoS turned on, because of the added load redirects introduce. It does however show higher RPS (and mean latency), because rendering a login page is less intensive than the issue search.

This test result doesn't reflect real user scenarios, can't be used prove that "redirection can't be used".

There are just two requests for `QoS On (Redirect)` in your test so you see `p95 latency` becomes higher, but the "/user/login" page is quite fast and it won't be affected by crawlers. And when the crawlers see the 503 or user login page, it would just stop that traversal because there is no more valid targets for it.

In a real world, when your server load is high, more than 99% CPU is for rendering the "expensive" pages, responding 503 or "/user/login" has almost no difference.

wxiaoguang commented on custom/conf/app.example.ini:
```diff
@@ -923,6 +923,16 @@ LEVEL = Info
 ;VALID_SITE_URL_SCHEMES=http,https
+;;
+;; Enable request quality of service and load shedding.
+; QOS_ENABLED = false
```
I agree to make the config system better, actually I think we need to make it flexible enough, because I think sooner or later we need to introduce more config options to let site admins to choose

Quote: https://github.com/go-gitea/gitea/pull/33951#discussion_r2032324964

> And according some feedbacks from #33966 and https://github.com/go-gitea/gitea/issues/33407#issuecomment-2776671993 , IMO a complete design could be like this:
> 1. Anti-crawler policy: block expensive pages, apply QoS to expensive pages
> 2. Anti-crawler action: respond 503, redirect to login, show a JS page to do some proof-of-work or use some rate limit token (this one seems to provide best user experience).

Then site admins could choose a combination to satisfy their requirements. I still think "show a JS page to do some proof-of-work or rate limit token" is the best choice for most public sites according to the feedbacks (I don't run a public site, so need public site admins to help)

bohde replied:
>  Instead of polluting [service], what about using a custom section, i.e. [qos]?

Thanks for the feedback! I put this under `[service.qos]` in https://github.com/go-gitea/gitea/pull/33951/commits/74fad6e6990c8122c20b1dd1c4271a52824484de. I had initially put it in the `[service]` config because I think it's related to the captcha settings, which are similarly used to address bots. Does this work for you?

wxiaoguang replied:
According to the "ini" package's buggy behavior, we should avoid using the "sub section"

Because all keys in its parent section will pollute the sub section, and cause hidden bugs.

So a clear top-level section should be the best choice.

More details are here:

![image](https://github.com/user-attachments/assets/786e5275-0476-4a8f-86a0-321f26b151a1)

bohde replied:
Thanks, I updated it to use `[qos]` in https://github.com/go-gitea/gitea/pull/33951/commits/e386e529bf944765bb870edc124b43f1809fbf17

bohde commented:
> That's not true according to the feedbacks of the existing `"REQUIRE_SIGNIN_VIEW=expensive" config option. All users said that there is no more CPU load.

The context for this statement is quite different from mine. `"REQUIRE_SIGNIN_VIEW=expensive"` makes all logged out requests to expensive pages redirect to a login page. This changes the work necessary to render one of these pages to performing a check if the user is logged in, and then serving the redirect. As long as this work is less than the work it would take otherwise render the page directly, this is a net win for every request. For example, if a server is serving 10 RPS to an endpoint, and an expensive page takes 500ms of CPU time to render, but the the login page render takes 100ms to render, this algorithm saves 10 * (500ms - 100ms) of work every second, or 4 CPU cores of work in savings. This is potentially a huge win for users, and directly aligns with the feedback from users. 

This context does not apply to how this algorithm works though. This algorithm specifically tries to serve as much traffic as possible, while maintaining latency targets. To use the numbers above again, if we have 10 RPS to a 500ms endpoint, but only 2 cores, this algorithm would converge on 4 RPS of goodput, while rejecting 6 RPS. 

Adding the redirect changes that math. Still using those same numbers, those 6 RPS that were formerly rejected now each generate another 100ms of work, at a default priority, which is higher than the low priority of the request that was initially turned into a redirect. This another 600ms of CPU time that is prioritized, only leaving 1.4 CPU seconds left to handle expensive requests, or slightly less than 3 RPS of goodput. This causes a feedback loop as well, where the reduced goodput then causes more redirects, but it will eventually converge to a lower amount of goodput on the server than if you served a 503. 

I do feel that this is turning into bikeshedding. We've been running this algorithm in production for a couple of months now, and I can confirm that it significantly helps deal with crawlers that do not respect our robots.txt, while still allowing crawlers that do to use our site. Suggestions like redirecting to login are not something I've tested in a production environment, so I can't recommend that approach. There are certainly other steps that can be done to address crawlers, and some of them such as `"REQUIRE_SIGNIN_VIEW=expensive"` are complementary approaches.

wxiaoguang commented:
> I do feel that this is turning into bikeshedding. 

I do not see any bikeshedding, I just suggested to make the design complete and flexible. I didn't say any "no" to your solution

> We've been running this algorithm in production for a couple of months now, and I can confirm that it significantly helps deal with crawlers

I agree, but it never answers the question above: "why a real anonymous end users should see a 503 error when the site is heavily being crawled?" My suggestion is to let site admin decide, you could still use 503 if you like, others could use redirection if they like.

wxiaoguang commented:
To be clear: if you have no objection, we can merge your solution as-is, and I will complete it and make it configurable to make site admin could choose the behavior they need.

bohde commented:
> To be clear: if you have no objection, we can merge your solution as-is, and I will complete it and make it configurable to make site admin could choose the behavior they need.

I think this solution is complete as is, and adding a redirect will have complex and unintended consequences to the algorithm that I detailed in https://github.com/go-gitea/gitea/pull/33951#issuecomment-2791241138. In previous production systems I've worked on, including those that used the library this is based on, I've seen slight tweaks to behavior such as the proposed redirect behavior cause feedback loops, which in turn cause worse outages than if the solution hadn't been in place. I always prefer to simplify the behavior in order to make it easy to reason about, and only adjust based upon production experience.

wxiaoguang commented:
I do not know why you have objection to let site admin choose the behavior they want.


But actually I am sure:

1. You could still use the 503 behavior by my proposal.
2. You haven't tried the user-redirection behavior by my proposal in Gitea. I am pretty sure the result should be the same as your 503 behavior. You can try it in your production to see if I was wrong. If I am wrong, I am happy to accept the conclusion and learn why.
3. You haven't tried the Proof-of-Work or rate-limit-token behavior by my proposal. I am pretty sure it will provide the best user experience among these behaviors.

bohde commented:
> You haven't tried the user-redirection behavior by my proposal in Gitea. I am pretty sure the result should be the same as your 503 behavior (CPU load). You can try it in your production to see if I was wrong. If I am wrong, I am happy to accept the conclusion and learn why.

I haven't seen your proposed code, but I have prototyped a version of it by serving a redirect to the login page and testing the behavior in local benchmarks in https://github.com/go-gitea/gitea/pull/33951#issuecomment-2790383514, and it is worse than the 503 method, because it now needs to also handle the login page request. These results are inline with my intuition of how it works in https://github.com/go-gitea/gitea/pull/33951#issuecomment-2791241138. Ultimately, I can only ever see this introducing a regression in both performance and latency in a production environment because it strictly adds work that a server needs to do. 

If I understand you, you think the added redirect will provide a nicer user experience, while the additional load of the login page will not meaningfully have a regression. Because this algorithm only ever rejects traffic when the server is already overloaded, any load added at this point, even if it is a small, makes this overloaded status worse, and causes more traffic to be rejected. This is a similar, but slightly different, failure mode as a retry storm. 

> You haven't tried the Proof-of-Work or rate-limit-token behavior by my proposal. I am pretty sure it will provide the best user experience among these behaviors. 

I'd certainly look at these, but I don't the goals of those are necessarily relevant to this PR. My goals in this PR is to keep friction low for anonymous users and crawlers that respect our robots.txt, which is why it only drops traffic if it needs to.

bohde commented:
> If you have no objection, I will propose one.

If you're fine merging this one, I would look at a follow up. 

> I think I could propose a "Proof-of-Work or rate-limit-token" solution and you could try it in your production to see whether it is better.

I would however need to see the proposed code before I could determine if I could test this in a production environment. Either way, I'm happy to take a look.

wxiaoguang commented:
> > If you have no objection, I will propose one.
> 
> If you're fine merging this one, I would look at a follow up.

Just like I said: I have no objection to this, public site admin maintainers could help, maybe: @techknowlogick @lunny

wxiaoguang commented:
> I would however need to see the proposed code before I could determine if I could test this in a production environment. Either way, I'm happy to take a look.

OK, I will propose one (maybe replace this one while achieve the same or better result)

bohde commented:
> > I would however need to see the proposed code before I could determine if I could test this in a production environment. Either way, I'm happy to take a look.
> 
> OK, I will propose one (maybe replace this one while achieve the same or better result)

I would need to see it of course, but I don't think the goals described in https://github.com/go-gitea/gitea/issues/33966#issuecomment-2785893126 are the same as this one. I think they may be complementary, but in our case we have public repos that we want both indexed by search engines and to be accessible by users who are not logged in, and want to keep friction as low as possible for those cases.  

I would prefer to merge this as is, and address any others in a follow up.

wxiaoguang commented:
->  Rate-limit anonymous accesses to expensive pages when the server is overloaded #34167 

I think this one covers all the cases.

bohde commented:
> -> Rate-limit anonymous accesses to expensive pages when the server is overloaded #34167
> 
> I think this one covers all the cases.

The algorithm there is clearly a derivative of the one I am proposing here, but it rejects more requests in practice since it does not enqueue requests. It also takes part of this PR, and introduces it in a new one, while stripping my authorship. This is evident in details such as the [default config setting](https://github.com/go-gitea/gitea/pull/34167/files#diff-0ab2a298fe8b3109330b3679b3ae75683af157d2e1123251f21e4bf6ed0c08efR170), which is copy pasted [from this PR](https://github.com/go-gitea/gitea/pull/33951/files#diff-0ab2a298fe8b3109330b3679b3ae75683af157d2e1123251f21e4bf6ed0c08efR292), but really only makes sense in the context of this algorithm. 

I've done a lot of legwork on this PR, including testing, benchmarking and refining it on production crawler traffic. I'm always open to feedback and ideas on how to improve an implementation, but have found this process very frustrating, from the [initial dismissive response](https://github.com/go-gitea/gitea/pull/33951#issuecomment-2741943250), the ongoing bikeshedding, and now the rewrite while stripping my authorship.

wxiaoguang commented:
> The algorithm there is clearly a derivative of the one I am proposing here, but it rejects more requests in practice since it does not enqueue requests.

I have explained many times: production speaks. You can't say that it is worse without testing it in real production environment. The crawler's behavior is different for your local testing.

> It also takes part of this PR, and introduces it in a new one, while stripping my authorship. This is evident in details such as the [default config setting](https://github.com/go-gitea/gitea/pull/34167/files#diff-0ab2a298fe8b3109330b3679b3ae75683af157d2e1123251f21e4bf6ed0c08efR170), which is copy pasted [from this PR](https://github.com/go-gitea/gitea/pull/33951/files#diff-0ab2a298fe8b3109330b3679b3ae75683af157d2e1123251f21e4bf6ed0c08efR292), but really only makes sense in the context of this algorithm.

I don't understand, if you are talking about "authorship", you copied my `routePattern := rctx.RoutePattern()` and other related code, where is the authorship? Please show me what to do, I will follow you and respect your authorship.

And by the way, "really only makes sense in the context of this algorithm" it is not true, I have explained in that PR's description. The "queue" work could be done by the browser and crawler.


> I've done a lot of legwork on this PR, including testing, benchmarking and refining it on production crawler traffic. I'm always open to feedback and ideas on how to improve an implementation, but have found this process very frustrating, from the [initial dismissive response](https://github.com/go-gitea/gitea/pull/33951#issuecomment-2741943250), the ongoing bikeshedding, and now the rewrite while stripping my authorship.

I also found that it is frustrating to fix various problems caused by various design problems (more than a half of these these 1400 PRs https://github.com/go-gitea/gitea/pulls/wxiaoguang are about "fixing this", "fixing that")

I have asked many times that "showing 503 blank error page is not friendly to end users", but no reasonable response. You just kept ignoring this fact.

Again, for the "authorship", please show me that how to respect the authorship for your copy of my `routePattern := rctx.RoutePattern()` and then I will follow it.

Thank you very much.

wxiaoguang commented:
By the way,  that PR #34167 is only an initial work, there are still some TODOs, and if it looks good, you could still integrate the `codel` package to use the builtin queue to defer some requests. I didn't do that because I don't want make it more complex at first stagen, and I know the `codel` work is from you, and I respect your authorship and decided to leave that improvement to you.

bohde commented:
> By the way, that PR #34167 is only an initial work, there are still some TODOs, and if it looks good, you could still integrate the `codel` package to use the builtin queue (Controlled Delay algorithm) to delay some requests. I didn't do that because I don't want make it more complex at first stage, and I know the `codel` work is from you, and I respect your authorship and decided to leave that improvement to you.

If that is the intent, then that PR should follow this one. This implements general overload protection, not just crawler mitigations. For example, if a logged in user performed a crawl, overload protection would still kick in, and keep latency down for all logged in users. This PR aims to solve a more general problem of keeping the server from having an outage. 

 This implementation does complement other crawler mitigation techniques, but they would likely benefit from extensions to this middleware.

bohde commented on routers/common/qos.go:
```diff
@@ -0,0 +1,112 @@
+// anonymous users are deprioritized.
+func QoS() func(next http.Handler) http.Handler {
+	if !setting.Service.QoS.Enabled {
+		return nil
```
When middlewares are added, nil handlers are filtered out at at https://github.com/go-gitea/gitea/blob/02e49a0f471fcd50e70835458d196615f03c39cc/modules/web/router.go#L98. 

This appears to be the preferred style according to other similar settings: https://github.com/go-gitea/gitea/blob/02e49a0f471fcd50e70835458d196615f03c39cc/routers/common/blockexpensive.go#L19

wxiaoguang commented:
I have made a PR for you: https://github.com/bohde/gitea/pull/1 , there was just no response in 2 weeks, but users were waiting. TBH I have no interest to waste time on inactive or stale PRs. At least, in history [I had to help to improve some of your PRs or propose some following up fixes for your PRs](https://github.com/go-gitea/gitea/pulls?q=is%3Apr+author%3Abohde+is%3Aclosed). A PR with "Maintainers are allowed to edit this pull request" should be open to maintainers to edit. I don't understand why you complain this time.

wxiaoguang commented:
> with rewritten commits from this PR, while not maintaining the commit history and thus my authorship.

That's not a rewritten, I write everyline from scratch. Hopefully you can understand. So there is no commit history.

Again, as I said, your PR (this PR) DIDN'T include any authorship of me, EITHER. So why complain? You just added some new commits later after you complained.

wxiaoguang commented:
And, one more thing, I just found this your new commit: ` if the client requests HTML, render an HTML 503 Service Unavailable page` : https://github.com/go-gitea/gitea/pull/33951/commits/5078f67e5b2384b4e9142cafa45f7a4d18f36394 , it is almost my "code" to render the template from another PR: if you complain I copied your code, to be honest : have your new commit included my authorship? Could I say or complain that you copied my code?

bohde commented:
Asking @go-gitea/maintainers to look at this. I think it has addressed all feedback, and other potential proposals do not address the same use cases as this. This is a general purpose load shedder that will help prevent a server from crashing in any overload scenario, that prioritizes traffic from logged users and deprioritizes traffic that is likely from a crawler. While it certainly helps a server that experiences significant crawler traffic, it also protects servers in a general denial of service cases, either from heavy usage from logged in users or targeted attacks. 

I've been running versions of this a production public server for a couple months now, and it has significantly improved the experience for logged in users.

The conversation has gotten quite long at this point, so I'll do my best to summarize the feedback and the status of it:

* ✅  [Use chi's RoutePattern to detect expensive paths and add tests](https://github.com/go-gitea/gitea/pull/33951#issuecomment-2741985551) was fixed in https://github.com/go-gitea/gitea/pull/33951/commits/aec549db53676988ec40d0aac55bd75cc62bb0a2
* ✅ [Serve a custom 503 page](https://github.com/go-gitea/gitea/pull/33951#issuecomment-2791581085) was addressed in https://github.com/go-gitea/gitea/pull/33951/commits/5078f67e5b2384b4e9142cafa45f7a4d18f36394
* [Use HTTP 429 instead of 503](https://github.com/go-gitea/gitea/pull/33951#issuecomment-2790472011) is not implemented for reasons [here](https://github.com/go-gitea/gitea/pull/33951#issuecomment-2790507737).

denyskon reviewed (approved):
I've read through this PR and the conversation, and I think this feature is more than complete enough for the goal it's trying to achieve. Further refinements could be done later, but it's crucial that we can ship this in 1.24, as many servers, including gitea.com, experience problems with high load recently.

lunny commented on routers/common/qos.go:
```diff
@@ -0,0 +1,144 @@
+			// Check if the request can begin processing.
+			err := c.Acquire(ctx, int(priority))
+			if err != nil {
+				renderServiceUnavailable(w, req)
```
Should we log the error here?

bohde replied:
This is a good callout! This was why I added the `.String()` method on `Priority`, but it was lost when I ported our internal implementation.

bohde commented:
Documentation PR is at https://gitea.com/gitea/docs/pulls/208, and the app.example.ini is updated to match in https://github.com/go-gitea/gitea/pull/33951/commits/6fb00219ae5be2af02fe8f5dacea123b59581e5d

delvh commented:
Since this PR seems to have some discussion but is ready otherwise, let's wait a day to see if anything changes.
If there's none, I'll merge this PR tomorrow.
